### ISSUE TITLE
Ignore patterns combining `/**/` and `{a,b}` alternatives not matching as expected

### ISSUE THREAD

**Issue by tomasz1986** [syncthing/syncthing#8733]
For reference: https://forum.syncthing.net/t/confusing-behaviour-when-trying-to-ignore-direct-and-indirect-child-files-involving-curly-bracketed-alternatives/19536

1. Create the following folder structure

```
Documents/document.txt
Documents/Photos/document.txt
```

2. Add `Documents/**/document.txt` to ignore patterns. Both files have been ignored.
3. Replace the patterns with `Documents/**/{document.txt}`. Only `Documents/Photos/document.txt` has been ignored.
4. Replace the patterns with `Documents{/**/}{document.txt}`. Now both files have been ignored again.

The behaviour doesn't seem consistent because a) both files should be ignored in 3), and b) there should be no difference between 3) and 4) while in reality the two patterns behave differently.

For the record, tested under Windows 10 and Syncthing v1.22.2.

**Comment by simonsimon006** [comment]
This seems like an inconsistency with the glob package instead of syncthing itself. E.g. when one uses the following code:
```
package main

import (
    "fmt"
    "github.com/gobwas/glob"
)

func main() {

    g := glob.MustCompile("api.**.com", '.')

    fmt.Printf("%s \n",  g.Match("api.github.com")) // true // 1
    fmt.Printf("%s \n", g.Match("api......com")) // true // 2
    fmt.Printf("%s \n",  g.Match("api.com")) // true but should be false // 3

    g2 := glob.MustCompile("api.**.{com}", '.')

    fmt.Printf("%s \n",  g2.Match("api.git.hub.com")) // true // 4
    fmt.Printf("%s \n",  g2.Match("api.com")) // false // 5

}
```
statement 3 should be false because two dots in the path should be necessary for the path to match. When the second pattern is used the (probably) correct behavior applies for some reason. Maybe this is some kind of user friendliness thing from glob. But without an exact description of what the "Super-Wildcard" is supposed to do exactly it cannot be decided.

Also, there is an issue for the glob package with a similar problem, see [this](https://github.com/gobwas/glob/issues/58) .

**Comment by calmh** [comment]
Yes, I came to the same conclusion (but apparently didn't bother to write it down).

**Comment by a-kbd** [comment]
One could rewrite the ignore functionality to use `doublestar` instead of the `glob` package, it seems to not have the issue described above (see code below), however, I'm unsure if it allows the possibility of compiling the globs and what implication that would have on performance. @calmh, we're a group currently in a course in KTH uni asking us to solve an issue in an open source project, would such a rewrite to use doublestar be reasonable/desirable in your opinion? In that case we'll be happy to work on it :)

```go
	g := glob.MustCompile("api.**.com", '.')

	fmt.Printf("%s \n", g.Match("api.github.com")) // true
	fmt.Printf("%s \n", g.Match("api......com"))   // true
	fmt.Printf("%s \n", g.Match("api.com"))        // true but should be false

	g2 := glob.MustCompile("api.**.{com}", '.')

	fmt.Printf("%s \n", g2.Match("api.git.hub.com")) // true
	fmt.Printf("%s \n", g2.Match("api.com"))         // false

	d, _ := doublestar.Match("api.**.com", "api.github.com")
	fmt.Printf("%v\n", d) // true

	d, _ = doublestar.Match("api.**.com", "api......com")
	fmt.Printf("%v\n", d) // true

	d, _ = doublestar.Match("api.**.com", "api.com")
	fmt.Printf("%v\n", d) // #### False as expected ####

	d, _ = doublestar.Match("api.**.{com}", "api.git.hub.com")
	fmt.Printf("%v\n", d) // true

	d, _ = doublestar.Match("api.**.{com}", "api.com")
	fmt.Printf("%v\n", d) // false
	```

**Comment by calmh** [comment]
@a-kbd I think it would be, if it adheres to the current meaning of the patterns (I think we have decent test coverage on those) and isn't too horrible a performance regression.

**Comment by a-kbd** [comment]
@calmh We have an implementation now, however, we are unsure about how to benchmark the performance properly. Is there a good way of testing the performance of only the file discovery/ignoring?

**Comment by calmh** [comment]
```
% go test -bench '^BenchmarkMatch$' -run none
goos: darwin
goarch: arm64
pkg: github.com/syncthing/syncthing/lib/ignore
BenchmarkMatch-20    	 2611910	       449.5 ns/op
PASS
ok  	github.com/syncthing/syncthing/lib/ignore	1.793s
```

**Comment by calmh** [comment]
That's within the margin of error, I see no problem. If it were a magnitude or more slower I would have wanted some attempt to mitigate it.

**Comment by imsodin** [comment]
Did you look at the cached benchmark, possibly? The other one is roughly a factor 4 slower. Not quite a magnitude, but significant. I see you already opened a ticket to improve on this upstream: https://github.com/bmatcuk/doublestar/issues/85  
It's hard to judge these things, because some of our users seem to be extremely performance sensitive. It's still very fast, and shouldn't have a big impact on our median usage (4.4k files, assuming 100 patterns that's 4.4e3 * 1e2 * 4e-6s ~= 1.6s (reality will differ of course)). At the same time I get flash-fowards to someone with a 99percentile usage on a tiny setup getting slowed down significantly :)  
Lets maybe wait for upstreams reaction on that ticket, but generally I am inclined to favor correctness and active maintainership over performance.

**Comment by calmh** [comment]
I did look at the cached values, yes, sorry. Honestly I think it's still fine, if it can be improved that's good, but otherwise I think time from ignore matching is already a very small fraction of the total. And there's always caching which can be enabled to shift the cost from cpu to ram.

**Comment by a-kbd** [comment]
Sorry for the complete radio silence for a while guys, we had a very high workload during our exam period. We made a pull request now, there are some comments included in our changes

**Comment by tomasz1986** [comment]
I've just had an issue with `{folder1/,folder2/}` not working at all (i.e. not ignoring anything), but `{folder1,folder2}/` does work. Is this related to this problem or is this something else?

**Comment by calmh** [comment]
Probably related. The globbing library we use have various funky corner cases and bugs for any advanced patterns; you can see further in their bug tracker.

Unfortunately switching it out as was started on isn't trivial, I tried to finish it but there were issues with something, backslashes IIRC.
